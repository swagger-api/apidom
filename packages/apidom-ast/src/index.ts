// JSON AST related exports
export { default as JsonNode } from './json/nodes/JsonNode.ts';
export { default as JsonDocument } from './json/nodes/JsonDocument.ts';
export { default as JsonObject } from './json/nodes/JsonObject.ts';
export { default as JsonProperty } from './json/nodes/JsonProperty.ts';
export type { JsonValue as JsonPropertyValue } from './json/nodes/JsonProperty.ts';
export { default as JsonArray } from './json/nodes/JsonArray.ts';
export { default as JsonValue } from './json/nodes/JsonValue.ts';
export type { JsonValueOptions } from './json/nodes/JsonValue.ts';
export { default as JsonKey } from './json/nodes/JsonKey.ts';
export { default as JsonString } from './json/nodes/JsonString.ts';
export { default as JsonStringContent } from './json/nodes/JsonStringContent.ts';
export { default as JsonEscapeSequence } from './json/nodes/JsonEscapeSequence.ts';
export { default as JsonNumber } from './json/nodes/JsonNumber.ts';
export { default as JsonTrue } from './json/nodes/JsonTrue.ts';
export { default as JsonFalse } from './json/nodes/JsonFalse.ts';
export { default as JsonNull } from './json/nodes/JsonNull.ts';
export {
  isDocument as isJsonDocument,
  isFalse as isJsonFalse,
  isProperty as isJsonProperty,
  isStringContent as isJsonStringContent,
  isEscapeSequence as isJsonEscapeSequence,
  isArray as isJsonArray,
  isKey as isJsonKey,
  isNull as isJsonNull,
  isNumber as isJsonNumber,
  isObject as isJsonObject,
  isString as isJsonString,
  isTrue as isJsonTrue,
} from './json/nodes/predicates.ts';
// YAML AST related exports
export { default as YamlAlias } from './yaml/nodes/YamlAlias.ts';
export type { YamlAliasOptions } from './yaml/nodes/YamlAlias.ts';
export { default as YamlCollection } from './yaml/nodes/YamlCollection.ts';
export { default as YamlComment } from './yaml/nodes/YamlComment.ts';
export type { YamlCommentOptions } from './yaml/nodes/YamlComment.ts';
export { default as YamlDirective } from './yaml/nodes/YamlDirective.ts';
export type { YamlDirectiveOptions, YamlDirectiveParameters } from './yaml/nodes/YamlDirective.ts';
export { default as YamlDocument } from './yaml/nodes/YamlDocument.ts';
export { default as YamlKeyValuePair } from './yaml/nodes/YamlKeyValuePair.ts';
export type { YamlKeyValuePairOptions } from './yaml/nodes/YamlKeyValuePair.ts';
export { default as YamlMapping } from './yaml/nodes/YamlMapping.ts';
export { default as YamlNode } from './yaml/nodes/YamlNode.ts';
export type { YamlNodeOptions } from './yaml/nodes/YamlNode.ts';
export { default as YamlScalar } from './yaml/nodes/YamlScalar.ts';
export type { YamlScalarOptions } from './yaml/nodes/YamlScalar.ts';
export { default as YamlSequence } from './yaml/nodes/YamlSequence.ts';
export { default as YamlStream } from './yaml/nodes/YamlStream.ts';
export { default as YamlTag, YamlNodeKind } from './yaml/nodes/YamlTag.ts';
export type { YamlTagOptions } from './yaml/nodes/YamlTag.ts';
export { default as YamlAnchor } from './yaml/nodes/YamlAnchor.ts';
export type { YamlAnchorOptions } from './yaml/nodes/YamlAnchor.ts';
export { YamlStyle, YamlStyleGroup } from './yaml/nodes/YamlStyle.ts';
export { default as YamlFailsafeSchema } from './yaml/schemas/failsafe/index.ts';
export { default as YamlJsonSchema } from './yaml/schemas/json/index.ts';
export { default as YamlReferenceManager } from './yaml/anchors-aliases/ReferenceManager.ts';
export {
  isAlias as isYamlAlias,
  isAnchor as isYamlAnchor,
  isComment as isYamlComment,
  isKeyValuePair as isYamlKeyValuePair,
  isDirective as isYamlDirective,
  isDocument as isYamlDocument,
  isMapping as isYamlMapping,
  isScalar as isYamlScalar,
  isSequence as isYamlSequence,
  isStream as isYamlStream,
  isTag as isYamlTag,
} from './yaml/nodes/predicates.ts';
export { default as YamlError } from './yaml/errors/YamlError.ts';
export { default as YamlReferenceError } from './yaml/errors/YamlReferenceError.ts';
export { default as YamlSchemaError } from './yaml/errors/YamlSchemaError.ts';
export { default as YamlTagError } from './yaml/errors/YamlTagError.ts';
export type { YamlTagErrorOptions } from './yaml/errors/YamlTagError.ts';
// generic AST related exports
export type { default as Node, NodeOptions } from './Node.ts';
export { default as Literal } from './Literal.ts';
export type { LiteralOptions } from './Literal.ts';
export { Point, default as Position } from './Position.ts';
export type { PointOptions, PositionOptions } from './Position.ts';
export { default as Error } from './Error.ts';
export type { ErrorOptions } from './Error.ts';
export { default as ParseResult } from './ParseResult.ts';
export { isParseResult, isLiteral, isPoint, isPosition } from './predicates.ts';
// AST traversal related exports
export {
  BREAK,
  mergeAll as mergeAllVisitors,
  getVisitFn,
  visit,
  getNodeType,
  isNode,
  cloneNode,
} from './traversal/visitor.ts';
export type { MergeAllSync, MergeAllAsync } from './traversal/visitor.ts';
