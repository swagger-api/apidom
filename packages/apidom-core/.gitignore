/dist
/es
/cjs
/types
/NOTICE
/swagger-api-apidom-core-*.tgz
