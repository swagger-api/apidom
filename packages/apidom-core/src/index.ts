import { has } from 'ramda';
import { isPlainObject, isString } from 'ramda-adjunct';
import { Element, Namespace as INamespace } from 'minim';

import defaultNamespaceInstance from './namespace';
import serializeValue from './serializers/value';
import serializeJSON from './serializers/json';
import serializeYAML from './serializers/yaml-1-2';

export { dispatchPlugins as dispatchRefractorPlugins } from './refractor/plugins/utils';
export { default as refractorPluginElementIdentity } from './refractor/plugins/element-identity';
export { default as refractorPluginSemanticElementIdentity } from './refractor/plugins/semantic-element-identity';

export { default as MediaTypes } from './media-types';

export { Element, MemberElement, KeyValuePair, ObjectSlice, ArraySlice, refract } from 'minim';
export type { NamespacePluginOptions, Attributes, Meta } from 'minim';
export type { PositionRange, Position } from './elements/SourceMap';
export { default as namespace, Namespace, createNamespace } from './namespace';

export {
  ObjectElement,
  ArrayElement,
  BooleanElement,
  NullElement,
  NumberElement,
  StringElement,
  LinkElement,
  RefElement,
  AnnotationElement,
  CommentElement,
  ParseResultElement,
  SourceMapElement,
} from './refractor/registration';

export {
  isElement,
  isStringElement,
  isNumberElement,
  isNullElement,
  isBooleanElement,
  isArrayElement,
  isObjectElement,
  isMemberElement,
  isLinkElement,
  isRefElement,
  isAnnotationElement,
  isParseResultElement,
  isSourceMapElement,
  isPrimitiveElement,
  hasElementSourceMap,
  includesSymbols,
  includesClasses,
} from './predicates/index';
export { default as createPredicate } from './predicates/helpers';

export { filter, reject, find, findAtOffset, some, traverse, parents } from './traversal/index';
export {
  visit,
  BREAK,
  mergeAllVisitors,
  getNodeType,
  keyMapDefault as keyMap,
} from './traversal/visitor';
export { transclude, default as Transcluder } from './transcluder/index';
export { dereference } from './util';
export { cloneShallow, cloneDeep } from './clone';
export { default as CloneError } from './clone/errors/CloneError';
export { default as DeepCloneError } from './clone/errors/DeepCloneError';
export { default as ShallowCloneError } from './clone/errors/ShallowCloneError';

/**
 * Transforms data to an Element from a particular namespace.
 */
export const from = (data: any, namespace: INamespace = defaultNamespaceInstance): Element => {
  if (isString(data)) {
    // JSON serialized refract
    try {
      return namespace.fromRefract(JSON.parse(data));
    } catch {
      // noop
    }
  }
  if (isPlainObject(data) && has('element', data)) {
    // refract javascript structure
    return namespace.fromRefract(data);
  }

  return namespace.toElement(data);
};

/**
 * Transforms the ApiDOM into JavaScript POJO.
 * This POJO would be the result of interpreting the ApiDOM
 * into JavaScript structure.
 */
export const toValue = serializeValue;

/**
 * Transforms the ApiDOM into JSON string.
 */
export const toJSON = serializeJSON;

/**
 * Transforms the ApiDOM into YAML string.
 */
export const toYAML = serializeYAML;

/**
 * Creates a refract representation of an Element.
 * https://github.com/refractproject/refract-spec
 */
export const dehydrate = (
  element: Element,
  namespace: INamespace = defaultNamespaceInstance,
): any => {
  return namespace.toRefract(element);
};

/**
 * Create a refracted string representation of an Element.
 */
export const toString = (
  element: Element,
  namespace: INamespace = defaultNamespaceInstance,
): string => {
  const refractStructure = dehydrate(element, namespace);
  return JSON.stringify(refractStructure);
};

export { default as sexprs } from './sexprs';
export { default as deepmerge } from './deepmerge';
