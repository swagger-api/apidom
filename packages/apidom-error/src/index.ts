// base error classes
export { default as ApiDOMError } from './ApiDOMError';
export { default as ApiDOMAggregateError } from './ApiDOMAggregateError';
export { default as ApiDOMStructuredError } from './ApiDOMStructuredError';
export type { default as ApiDOMErrorOptions } from './ApiDOMErrorOptions';
// generic custom error classes
export { default as UnsupportedOperationError } from './UnsupportedOperationError';
export { default as NotImplementedError } from './NotImplementedError';
