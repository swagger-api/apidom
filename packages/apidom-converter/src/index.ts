import { ParseResultElement } from '@swagger-api/apidom-core';
import { mergeOptions, bundle, File } from '@swagger-api/apidom-reference';

import defaultOptions, { ConverterOptions } from './options';
import ConvertError from './errors/ConvertError';
import UnmatchedConvertStrategyError from './errors/UnmatchedConvertStrategyError';

export { ConvertError, UnmatchedConvertStrategyError };

/**
 * `convertApiDOM` already assumes that the ApiDOM is bundled.
 */
export const convertApiDOM = async (element: ParseResultElement, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const file = File({
    uri: mergedOptions.resolve.baseURI,
    parseResult: element,
    mediaType: mergedOptions.convert.sourceMediaType || mergedOptions.parse.mediaType,
  });
  const strategy = mergedOptions.convert.strategies.find((s) => s.canConvert(file, mergedOptions));

  if (typeof strategy === 'undefined') {
    throw new UnmatchedConvertStrategyError(file.uri);
  }

  return strategy.convert(file, mergedOptions);
};

const convert = async (uri: string, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const parseResult = await bundle(uri, mergedOptions);

  return convertApiDOM(parseResult, mergedOptions);
};

export default convert;
