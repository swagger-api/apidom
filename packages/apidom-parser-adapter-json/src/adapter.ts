import { createNamespace } from '@swagger-api/apidom-core';

export { default as mediaTypes } from './media-types';

export const namespace = createNamespace();

export const detectionRegExp =
  /(^\s*true\s$)|(^\s*false\s*$)|(^\s*null\s*$)|(^\s*\d+\s*$)|(^\s*{\s*)|(^\s*\[\s*)|(^\s*"([^"\\\\]*|\\\\["\\\\bfnrt/]|\\\\u[0-9a-f]{4})*"\s*$)/;
