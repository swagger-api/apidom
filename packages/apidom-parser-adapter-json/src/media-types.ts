import { MediaTypes } from '@swagger-api/apidom-core';

export class JSONMediaTypes extends MediaTypes<string> {
  latest() {
    return this[0];
  }
}

const mediaTypes = new JSONMediaTypes('application/json');

export default mediaTypes;
