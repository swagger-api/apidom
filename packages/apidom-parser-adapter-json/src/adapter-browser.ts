import { ParseResultElement } from '@swagger-api/apidom-core';

import lexicalAnalysis from './lexical-analysis/browser.ts';
import syntacticAnalysisDirect from './syntactic-analysis/direct/index.ts';
import syntacticAnalysisIndirect from './syntactic-analysis/indirect/index.ts';
import { detectionRegExp } from './adapter.ts';

export { mediaTypes, namespace } from './adapter.ts';
export type { JSONMediaTypes } from './media-types.ts';
export type { Tree } from './syntactic-analysis/indirect/index.ts';
export { detectionRegExp };

export {
  lexicalAnalysis,
  syntacticAnalysisDirect as syntacticAnalysis,
  syntacticAnalysisDirect,
  syntacticAnalysisIndirect,
};

/**
 * @public
 */
export const detect = async (source: string): Promise<boolean> => {
  if (!detectionRegExp.test(source)) {
    return false;
  }

  try {
    const cst = await lexicalAnalysis(source);
    return cst.rootNode.type !== 'ERROR';
  } catch {
    return false;
  }
};

/**
 * @public
 */
export interface ParseFunctionOptions {
  sourceMap?: boolean;
  syntacticAnalysis?: 'direct' | 'indirect';
}

/**
 * @public
 */
export type ParseFunction = (
  source: string,
  options?: ParseFunctionOptions,
) => Promise<ParseResultElement>;

/**
 * @public
 */
export const parse: ParseFunction = async (
  source,
  { sourceMap = false, syntacticAnalysis = 'direct' } = {},
) => {
  const cst = await lexicalAnalysis(source);
  let apiDOM;

  if (syntacticAnalysis === 'indirect') {
    apiDOM = syntacticAnalysisIndirect(cst, { sourceMap });
  } else {
    apiDOM = syntacticAnalysisDirect(cst, { sourceMap });
  }

  return apiDOM;
};
