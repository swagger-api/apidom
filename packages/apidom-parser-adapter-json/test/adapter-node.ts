import fs from 'fs';
import path from 'path';
import { expect, assert } from 'chai';
import { sexprs, isObjectElement, isParseResultElement } from '@swagger-api/apidom-core';

import * as adapter from '../src/adapter-node';

const spec = fs.readFileSync(path.join(__dirname, 'fixtures', 'sample-data.json')).toString();

describe('adapter-node', function () {
  context('given valid JSON', function () {
    specify('should detect proper media type', async function () {
      assert.isTrue(await adapter.detect(spec));
    });
  });

  context('given non JSON', function () {
    specify('should detect proper media type', async function () {
      assert.isFalse(await adapter.detect('^}'));
    });
  });

  it('should parse', async function () {
    const parseResult = await adapter.parse(spec, {
      syntacticAnalysis: 'direct',
      sourceMap: true,
    });

    assert.isTrue(isParseResultElement(parseResult));
    assert.isTrue(isObjectElement(parseResult.result));
    expect(sexprs(parseResult)).toMatchSnapshot();
  });

  context('given direct syntactic analysis', function () {
    context('given zero byte empty file', function () {
      specify('should return empty parse result', async function () {
        const parseResult = await adapter.parse('', {
          sourceMap: true,
          syntacticAnalysis: 'direct',
        });

        assert.isTrue(parseResult.isEmpty);
      });
    });

    context('given non-zero byte empty file', function () {
      specify('should return empty parser result', async function () {
        const parseResult = await adapter.parse('  ', {
          sourceMap: true,
          syntacticAnalysis: 'direct',
        });

        assert.isTrue(parseResult.isEmpty);
      });
    });

    context('given invalid json file', function () {
      specify('should return empty parser result', async function () {
        const parseResult = await adapter.parse(' a ', {
          sourceMap: true,
          syntacticAnalysis: 'direct',
        });

        assert.isTrue(parseResult.isEmpty);
      });
    });
  });

  context('given indirect syntactic analysis', function () {
    context('given zero byte empty file', function () {
      specify('should return empty parse result', async function () {
        const parseResult = await adapter.parse('', {
          sourceMap: true,
          syntacticAnalysis: 'indirect',
        });

        assert.isTrue(parseResult.isEmpty);
      });
    });

    context('given non-zero byte empty file', function () {
      specify('should return empty parser result', async function () {
        const parseResult = await adapter.parse('  ', {
          sourceMap: true,
          syntacticAnalysis: 'indirect',
        });

        assert.isTrue(parseResult.isEmpty);
      });
    });

    context('given invalid json file', function () {
      specify('should return empty parser result', async function () {
        const parseResult = await adapter.parse(' a ', {
          sourceMap: true,
          syntacticAnalysis: 'indirect',
        });

        assert.isTrue(parseResult.isEmpty);
      });
    });
  });
});
