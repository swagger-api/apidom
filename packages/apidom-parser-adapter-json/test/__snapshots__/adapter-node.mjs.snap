// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter-node should parse 1`] = `
(ParseResultElement
  (ObjectElement
    (MemberElement
      (StringElement)
      (StringElement))
    (MemberElement
      (StringElement)
      (BooleanElement))
    (MemberElement
      (StringElement)
      (StringElement))))
`;
